src/data/equiv/one_sided.lean

```diff
@@ -0,0 +1,215 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+import data.set.function
+
+/-!
+# Functions with a bundled one-sided inverse
+
+This provides `α ↬ β` (`injection`) and `α ↠ β` (`surjection`), for functions associated with a
+specific left- or right-inverse. For the two-sided inverse, use `α ≃ β` (`equiv`).
+
+For cases when you care only that a function is injective, and do not need a specific inverse, use
+`α ↪ β` (`embedding`).
+-/
+
+open function
+
+universes u v w z
+variables {α : Sort u} {β : Sort v} {γ : Sort w} {δ : Sort w}
+
+set_option old_structure_cmd true
+
+/-- `α ↬ β` is the type of functions from `α → β` with an associated left-inverse. -/
+@[nolint has_inhabited_instance]
+structure injection (α : Sort*) (β : Sort*) :=
```

Comment:
I'm unsure about the use of the name `injection` and `surjection` here - I'd expect an injection to mean a function together with a proof that it's injective (ie what's already called an embedding), and a surjection to mean a function with a proof that it's surjective; but the meaning here is that an injection is a function with its left inverse and a surjection is a function with its right inverse, which is a different notion. Specifically not every injective function has a left inverse (and the claim that every surjective function has a right inverse is equivalent to choice). I actually prefer your original names `left_invertible` and `right_invertible` to these, though perhaps better names are available.

Reply:
Yeah, I don't like these names either

Reply:
Yeah, `injection` is not correct here. I think `split_mono` or better `split_injection` (which hopefully makes it clearer this is just a `Type`-level notion).